### PR TITLE
Add support for 9-sliced backgrounds

### PR DESCRIPTION
**Pull request by rubenwardy**
This PR adds support for 9-sliced textures to Minetest. 9-slice textures are commonly used in GUIs to allow scaling them to match any resolution without distortion.

This PR only implements 9-slices to be used with the `background[]` element.

image

https://en.wikipedia.org/wiki/9-slice_scaling

## To do

- [ ] Think of better place to put `draw2DImage9Splice`

## How to test

[](https://i.rubenwardy.com/Iltif.png)

```lua
minetest.register_on_joinplayer(function(player)
    minetest.show_formspec(player:get_player_name(), "aaa", [[
            size[4,6]
            background[5,5;1,1;testbg.png;true;10]
       ]])
end)
```

### CODE REVIEW

File: src/client/guiscalingfilter.cpp [hunk 1]
```diff
@@ -167,3 +167,49 @@ void draw2DImageFilterScaled(video::IVideoDriver *driver, video::ITexture *txr,
...
+	const video::SColor colors[] = {color,color,color,color};
+
+	auto originalSize = texture->getOriginalSize();
+	core::vector2di lowerRightOffset = core::vector2di(originalSize.Width, originalSize.Height) - middle.LowerRightCorner;
+
+	for (int y = 0; y < 3; ++y) {
+		for (int x = 0; x < 3; ++x) {
+			core::rect<s32> src({0, 0}, originalSize);
+			core::rect<s32> dest = rect;
+			if (x == 0) {
```
Comment (nerzhul) [inline src/client/guiscalingfilter.cpp]: switch case please

File: src/client/guiscalingfilter.cpp [hunk 2]
```diff
@@ -167,3 +167,49 @@ void draw2DImageFilterScaled(video::IVideoDriver *driver, video::ITexture *txr,
...
+				dest.UpperLeftCorner.X += middle.UpperLeftCorner.X;
+				dest.LowerRightCorner.X -= lowerRightOffset.X;
+				src.UpperLeftCorner.X = middle.UpperLeftCorner.X;
+				src.LowerRightCorner.X = middle.LowerRightCorner.X;
+			} else if (x == 2) {
+				dest.UpperLeftCorner.X = rect.LowerRightCorner.X - lowerRightOffset.X;
+				src.UpperLeftCorner.X = middle.LowerRightCorner.X;
+			}
+
+			if (y == 0) {
```
Comment (nerzhul) [inline src/client/guiscalingfilter.cpp]: same here

File: src/gui/guiFormSpecMenu.cpp [hunk 3]
```diff
@@ -672,16 +672,37 @@ void GUIFormSpecMenu::parseBackground(parserData* data, const std::string &eleme
...
 		}
 
+		core::rect<s32> middle;
+		if (parts.size() >= 5) {
+			std::vector<std::string> v_middle = split(parts[4], ',');
+			if (v_middle.size() == 2) {
+				s32 x = stoi(v_middle[0]);
+				middle.UpperLeftCorner = core::vector2di(x, x);
+				middle.LowerRightCorner = core::vector2di(-x, -x);
+			} else if (v_middle.size() == 2) {
```
Comment (nerzhul) [inline src/gui/guiFormSpecMenu.cpp]: sounds to be same condition as upper

Comment (rubenwardy) [reply src/gui/guiFormSpecMenu.cpp]: ah, I forgot to push a typo fix

File: src/gui/guiFormSpecMenu.cpp [hunk 4]
```diff
@@ -653,7 +653,7 @@ void GUIFormSpecMenu::parseBackground(parserData* data, const std::string &eleme
 {
 	std::vector<std::string> parts = split(element,';');
 
-	if (((parts.size() == 3) || (parts.size() == 4)) ||
+	if (((parts.size() >= 3) || (parts.size() <= 5)) ||
 		((parts.size() > 4) && (m_formspec_version > FORMSPEC_API_VERSION)))
```
Comment (DS-Minetest) [inline src/gui/guiFormSpecMenu.cpp]:
```cpp
	if (((parts.size() >= 3) && (parts.size() <= 5)) ||
		((parts.size() > 5) && (m_formspec_version > FORMSPEC_API_VERSION)))
```

File: src/gui/guiFormSpecMenu.cpp [hunk 5]
```diff
@@ -2523,12 +2546,23 @@ void GUIFormSpecMenu::drawMenu()
 									AbsoluteRect.UpperLeftCorner.Y + absrec_size.Height + spec.pos.Y);
 			}
 
-			const video::SColor color(255,255,255,255);
-			const video::SColor colors[] = {color,color,color,color};
-			draw2DImageFilterScaled(driver, texture, rect,
-				core::rect<s32>(core::position2d<s32>(0,0),
-						core::dimension2di(texture->getOriginalSize())),
-				NULL/*&AbsoluteClippingRect*/, colors, true);
+			if (middle.getSize() == core::vector2di(0, 0)) {
```
Comment (DS-Minetest) [inline src/gui/guiFormSpecMenu.cpp]: Why not
```suggestion
			if (middle.getArea() == 0) {
```
?

File: src/gui/guiFormSpecMenu.cpp (hunk 4 above)
Comment (SmallJoker) [reply src/gui/guiFormSpecMenu.cpp]:
```C++
	if ((parts.size() >= 3 && parts.size() <= 5) ||
		(parts.size() > 5 && m_formspec_version > FORMSPEC_API_VERSION))
```
Annoying parentheses.

**Comment by SmallJoker** [comment]
The image resizing and formspec scaling (again) really sucks, but the PR does its job perfectly.

demo

File: src/gui/guiFormSpecMenu.cpp (hunk 5 above)
Comment (rubenwardy) [reply src/gui/guiFormSpecMenu.cpp]: I expect this could be more intensive. Maybe I should check to see if both sides are 0.

Comment (DS-Minetest) [reply src/gui/guiFormSpecMenu.cpp]: [`getSize`](http://irrlicht.sourceforge.net/docu/rect_8h_source.html#l00231) calls `getWidth` and `getHeight` and creates a new dimension, [`getArea`](http://irrlicht.sourceforge.net/docu/rect_8h_source.html#l00097) also calls `getWidth` and `getHeight` and just multiplies both.

File: doc/lua_api.txt [hunk 6]
```diff
@@ -2046,6 +2046,22 @@ Elements
...
+
+* 9-sliced background. See https://en.wikipedia.org/wiki/9-slice_scaling
+* Middle is a rect which defines the middle of the 9-slice.
+	* `x` - The middle will be x pixels from all sides.
+	* `x,y` - The middle will be x pixels from the horizontal and y from the vertical.
+	* `x,y,x2,y2` - The middle will start at x,y, and end at x2, y2. Negative x2 and y2 values
+		will be added to the width and height of the texture, allowing it to be used as the
+		distance from the far end.
+	* All numbers in middle are integers.
+* Use a background. Inventory rectangles are not drawn then.
```
Comment (SmallJoker) [inline doc/lua_api.txt]: What does this mean? Why are they not drawn?

Comment (rubenwardy) [reply doc/lua_api.txt]: This is copied from the old element's description, I'm not sure what it means

Comment (SmallJoker) [reply doc/lua_api.txt]: Remove it. I don't see any issue there.
grafik

**Pull request merged**